cmd/ndm_daemonset/probe/sysfsprobe.go

```diff
@@ -0,0 +1,162 @@
+*/
+
+/*
+
+Find queue sysfs doc here.
+https://www.kernel.org/doc/Documentation/block/queue-sysfs.txt
+
+*/
+
+package probe
+
+import (
+	"io/ioutil"
+	"strconv"
+	"strings"
+
+	"github.com/openebs/node-disk-manager/blockdevice"
+	"github.com/openebs/node-disk-manager/cmd/ndm_daemonset/controller"
+	"github.com/openebs/node-disk-manager/pkg/util"
+	"k8s.io/klog"
+)
+
+const (
+	sysfsProbePriority = 2
+	sysfsConfigKey     = "sysfs-probe"
+	// sectorSize is the sector size as understood by the unix systems.
+	// It is kept as 512 bytes. all entries in /sys/class/block/sda/size
+	// are in 512 byte blocks
+	sectorSize int64 = 512
+)
+
+var (
+	sysfsProbeName  = "sysfs probe"
+	sysfsProbeState = defaultEnabled
+)
+
+var sysfsProbeRegister = func() {
+	ctrl := <-controller.ControllerBroadcastChannel
+	if ctrl == nil {
+		klog.Error("unable to configure", sysfsProbeName)
+		return
+	}
+	if ctrl.NDMConfig != nil {
+		for _, probeConfig := range ctrl.NDMConfig.ProbeConfigs {
+			if probeConfig.Key == sysfsConfigKey {
+				sysfsProbeName = probeConfig.Name
+				sysfsProbeState = util.CheckTruthy(probeConfig.State)
+				break
+			}
+		}
+	}
+
+	newRegistryProbe := &registerProbe{
+		priority:   sysfsProbePriority,
+		name:       sysfsProbeName,
+		state:      sysfsProbeState,
+		pi:         newSysFSProbe(),
+		controller: ctrl,
+	}
+	newRegistryProbe.register()
+}
+
+// sysfsProbe fills the logical sector size,
+// physical sector size, drive type(ssd or hdd) of the disk
+type sysfsProbe struct {
+}
+
+func newSysFSProbe() *sysfsProbe {
+	return &sysfsProbe{}
+}
+
+// It is part of probe interface. Hence, empty implementation.
+func (cp *sysfsProbe) Start() {}
+
+// FillBlockDeviceDetails updates the logical sector size,
+// physical sector size, drive type(ssd or hdd) of the disk
+// if those are not populated.
+func (cp *sysfsProbe) FillBlockDeviceDetails(blockDevice *blockdevice.BlockDevice) {
+	sysPath := blockDevice.SysPath
+
+	if blockDevice.DeviceAttributes.LogicalBlockSize == 0 {
+		logicalBlockSize, err := readSysFSFileAsInt64(sysPath + "/queue/logical_block_size")
+		if err != nil {
+			klog.Warning("unable to read logical block size", err)
+		} else if logicalBlockSize != 0 {
+			blockDevice.DeviceAttributes.LogicalBlockSize = uint32(logicalBlockSize)
+			klog.Infof("blockdevice path: %s logical sector size :%d filled by sysfs probe.",
+				blockDevice.DevPath, blockDevice.DeviceAttributes.LogicalBlockSize)
+		}
+	}
+
+	if blockDevice.DeviceAttributes.PhysicalBlockSize == 0 {
+		physicalBlockSize, err := readSysFSFileAsInt64(sysPath + "/queue/physical_block_size")
+		if err != nil {
+			klog.Warning("unable to read physical block size", err)
+		} else if physicalBlockSize != 0 {
+			blockDevice.DeviceAttributes.PhysicalBlockSize = uint32(physicalBlockSize)
+			klog.Infof("blockdevice path: %s physical sector size :%d filled by sysfs probe.",
+				blockDevice.DevPath, blockDevice.DeviceAttributes.PhysicalBlockSize)
+		}
+	}
+
```

Comment:
There is a `HardwareSectorSize` field in `DeviceAttributes` which should be filled in sysfs-probe using `hw_sector_size`.

Reply:
ok.

Reply:
Done